### PR TITLE
Asynchronously load tinyMCE the first time a classic block is edited

### PR DESCRIPTION
**Pull request by sarayourfriend**
## Description

Added `LazyLoad` component. This component allows us to wrap existing components and delay them until required 3rd party scripts and styles are asynchronously loaded. It also provides a hook for post-load setup that resolves before the children are rendered, guaranteeing that the children aren't active until the world around them is set up correctly.

I've then wrapped `ClassicEdit` in `LazyLoad`. This 100% does not work and is just a proof of concept.

A related PR: https://github.com/WordPress/gutenberg/pull/21244

We'd replace `load-scripts/styles` in the current code here with requests to those endpoints.

## Types of changes
New feature

## Checklist:
- [ ] My code is tested.
- [ ] My code follows the WordPress code style. 
- [ ] My code follows the accessibility standards. 
- [ ] My code has proper inline documentation. 
- [ ] I've included developer documentation if appropriate. 
- [ ] I've updated all React Native files affected by any refactorings/renamings in this PR. 

## Testing instructions

1. Checkout https://github.com/WordPress/wordpress-develop/pull/232 locally and configure Gutenberg to point towards `wordpress-develop` by following the instructions for `wp-env` configuration here: https://github.com/WordPress/gutenberg/blob/master/packages/env/README.md#wp-envjson. Run WordPress using `npx wp-env start` inside your Gutenberg repo.
2. Checkout this PR locally and run `npm run dev`
3. Open the block editor and search the page source for the TinyMCE source code; ensure it does not exist. Additionally, check that `window.tinymce === undefined`
4. Edit a classic block and watch the network tab in your browser. Note that TinyMCE is loaded and `window.tinymce` is defined.
5. Ensure TinyMCE translations are loading correctly by changing your user's locale to one with different translations than what you're starting with and confirming that the translations worked.
6. Confirm that exiting a classic block and re-entering it (or another classic block) does _not_ make a new request for TinyMCE or the translations.

### CODE REVIEW

File: packages/block-library/src/classic/lazy.js [hunk 1]
```diff
@@ -0,0 +1,89 @@
...
+		);
+
+		if ( scriptsToLoad.length === 0 ) {
+			return Promise.resolve();
+		}
+
+		await new Promise( ( resolve, reject ) => {
+			const scriptElement = document.createElement( 'script' );
+			// can't use load-scripts.php, so this would need to be replaced
+			scriptElement.src = `/wp-admin/load-scripts.php?load=${ scriptsToLoad.join(
```
Comment (simison) [inline packages/block-library/src/classic/lazy.js:38]: I know this is a draft, but just a note to remember that `/wp-admin` can be in a subfolder; `example.org/sub/wp-admin`

I don't remember if it's possible to rename it but there's `admin_url` filter.

Comment (saramarcondes) [reply packages/block-library/src/classic/lazy.js:38]: Oh, thanks for pointing that out!

File: gutenberg.php [hunk 2]
```diff
@@ -159,3 +159,21 @@ function gutenberg_rest_nonce() {
...
+function gutenberg_print_tinymce_scripts() {
+	$current_screen = get_current_screen();
+	if ( ! $current_screen->is_block_editor() ) {
+		// maybe also need to check a setting/feature flag? Put this behind phase 2?
+		wp_print_tinymce_scripts();
+	} else {
+		echo "<!-- Skipping TinyMCE in favor of loading async -->";
+	}
+}
+remove_action( 'print_tinymce_scripts', 'wp_print_tinymce_scripts' );
```
Comment (saramarcondes) [inline gutenberg.php:178]: This action needs to be added in core, example PR here: https://github.com/WordPress/wordpress-develop/pull/232

File: packages/block-library/src/classic/edit.js [hunk 3]
```diff
@@ -216,3 +221,16 @@ export default class ClassicEdit extends Component {
 		/* eslint-enable jsx-a11y/no-static-element-interactions */
 	}
 }
+
+const LazyClassicEdit = ( props ) => (
+	<LazyLoad
+		scripts={ [ 'wp-tinymce' ] }
+		styles={ [ 'wp-tinymce' ] }
+		onLoaded={ () => window.wpMceTranslation() }
```
Comment (saramarcondes) [inline packages/block-library/src/classic/edit.js:229]: I'm not actually sure whether this would be how we do it. Given we can fully control what gets printed when Gutenberg is loading through the `print_tinymce_scripts` we can 100% wrap `wp_mce_translation`'s result in a function to run in `onLoaded`... but it's also been floated that along with the new REST API that will provide the TinyMCE script, we could return the info necessary for initializing the library's i18n as a JSON blob.

Right now I _think_ I prefer just printing this wrapped in a function in the action, but curious what the trade offs people can think of would be.

File: gutenberg.php [hunk 4]
```diff
@@ -159,3 +159,25 @@ function gutenberg_rest_nonce() {
...
+ */
+function gutenberg_print_tinymce_scripts() {
+	$current_screen = get_current_screen();
+	if ( ! $current_screen->is_block_editor() ) {
+		// maybe also need to check a setting/feature flag? Put this behind phase 2?
+		wp_print_tinymce_scripts();
+	} else {
+		echo "<!-- Skipping TinyMCE in favor of loading async -->";
+		echo "<script type='text/javascript'>\n" .
+			"window.wpMceTranslation = function() {\n" .
```
Comment (gziolo) [inline gutenberg.php:177]: Shouldn't it be part of the REST API response in this async scenario?

I don't think it can be as simple as that for the block editor. Plugins still can register [Meta Boxes](https://github.com/WordPress/gutenberg/blob/master/docs/designers-developers/developers/backward-compatibility/meta-box.md) that use TinyMCE heavily.

Comment (saramarcondes) [reply gutenberg.php:177]: Yeah. Definitely an open question to me... I guess we can create a filter? The issue is the endpoint returns just the URI for the dependency, not the dependency script itself, so we can’t just concatenate the translation bit to it.

An alternative to this is to just have a `tinymce-i18n` endpoint to just return the json needed, the async `onLoaded` prop is suitable for that.

File: packages/block-library/src/classic/edit.js [hunk 5]
```diff
@@ -225,12 +226,17 @@ class ClassicEdit extends Component {
...
+			const {
+				translations,
+				locale,
+				locale_script_handle: localeScriptHandle,
+			} = await apiFetch( { path: '/wp/v2/tinymce-i18n' } );
+
+			const { tinymce } = window;
+			tinymce.addI18n( locale, JSON.stringify( translations ) );
+			tinymce.ScriptLoader.markDone( localeScriptHandle );
+		} }
```
Comment (saramarcondes) [inline packages/block-library/src/classic/edit.js:239]: @gziolo I think these changes will be interesting to you 😊 

cc @griffbrad because we'd also discussed using an endpoint to retrieve the translation json.

File: lib/class-wp-rest-tinymce-i18n-controller.php [hunk 6]
```diff
@@ -0,0 +1,76 @@
...
+
+/**
+ * Endpoint for retrieving TinyMCE translations needed to initialize the library
+ * on the frontend. Used by the classic block's edit LazyLoad wrapper.
+ *
+ * Class WP_REST_TinyMCE_I18n_Controller
+ *
+ * @see WP_REST_Controller
+ */
+class WP_REST_TinyMCE_I18n_Controller extends WP_REST_Controller {
```
Comment (saramarcondes) [inline lib/class-wp-rest-tinymce-i18n-controller.php:16]: Still POCing, needs tests and an actually correct schema!

**Comment by azaozz** [comment]
I'm probably missing something (or a lot of things) but... Why is all the REST API stuff needed here? What is the purpose of it? Is it just to get a script's URL? Why would it need to use REST and then load the script(s) or stylesheet(s) without using REST? How is that better than loading them directly and not using REST at all?

If the current Script Loader in core needs to be used, it would be pretty simple to add a function that will get a script's handle as param, run wp, calculate the dependencies and output either the script tags plus the inline scripts, or the (concatenated) JS or CSS as a blob, or... anything else that makes sense.

In TinyMCE's case using REST imho seems even more "not-useful" as it doesn't have any dependencies and generally cannot be used as dependency. Getting the URL to `tinymce.min.js` would be as simple as doing `window.tinyMCEPreInit.baseURL + '/tinymce.min.js';`

**Comment by StevenDufresne** [comment]
I definitely share @azaozz's concern (and his missing of things 😊) that the `REST` calls to retrieve assets seem unnecessary seeing that we already know which blocks are registered and which assets are needed for those blocks on page load. 

Theoretically it may feel nice to reduce `ms` in page loading time, but depending on number of unique blocks, we may move to a place where the page loads a few `ms` faster but we've introduced a longer wait to interact because `x` number of blocks has queued up `x` number of http requests and the browser is throttling the requests (most noticeable in high latency contexts). 

With that being said, this PR is great because the `REST` call strategy is necessary when we don't know the assets ahead of time, as is the case with the block directory (although the block directory code can't be wrapped by a lazy load component at this moment in time). Anything learned here will be helpful for the block directory.

**Comment by aduth** [comment]
@azaozz @StevenDufresne If this was only ever intended to support TinyMCE, it might not be necessary to go to the extent of implementing a set of REST API endpoints. But assuming we'd want to apply this pattern to other blocks, or to [other parts of the editor screen](https://github.com/WordPress/gutenberg/issues/21820#issuecomment-620179173), I sense it's not something we can prepare ahead of time during the initial load. At least, I would think the full dependency tree of all scripts would be quite large and not something to send over the wire (worth verifying). Instead, I think we'd need to be able to compute on-the-fly if and when a script is needed to be loaded. As I see it, the endpoint largely encompasses what you've described as a function receiving a script handle, just that it's exposed via the REST API.

To the worry about delays from network requests: For the ones we can know ahead of time, there's already a pattern to preload REST API results at load time, so that the editor never needs to actually send a network request ([`rest_preload_api_request`](https://developer.wordpress.org/reference/functions/rest_preload_api_request/), [editor example](https://github.com/WordPress/wordpress-develop/blob/e46d1cfc84d3b4cd715ef07d3c3670cb4744048a/src/wp-admin/edit-form-blocks.php#L41-L78)).

**Comment by saramarcondes** [comment]
Yes, to reiterate what @aduth is saying here and in https://github.com/WordPress/gutenberg/issues/21738 (where I also wrote some more general comments about this being a generic approach to a specific solution... because we had to start somewhere and TinyMCE is a big win if we can get it), I don't want us to get bogged down in the weeds of TinyMCE's particularities.

RE: whether TinyMCE is too complicated a starting point for this... I feel strongly like _I_ must be missing something here given who all is skeptical of this... but the draft PR does indeed work, you can try it out if you run WP with https://github.com/WordPress/wordpress-develop/pull/232 to add the action that the Gutenberg PR uses to prevent TinyMCE from loading at the start. TinyMCE will not be loaded, but the second you edit a classic block, you'll see TinyMCE start to load and eventually the block works. I'll add more specific testing instructions to the PR to make it clear that this is a working POC.

RE: @azaozz concern about why a REST API is necessary: I'm in total agreement with @StevenDufresne here... the REST API helps out with the block directory is a non-trivial way. Below I propose a hybridzed approach based on @aduth's suggestion at the end of this comment https://github.com/WordPress/gutenberg/issues/21738#issuecomment-620611198. Hopefully this will help explicate why we do need a REST API while also addressing your concern about spinning up WP for too many REST requests. For now if only the classic block uses this, I don't think the concern about excessive requests applies yet, and below I've described a way we can proactively mitigate that problem by basically doing as you've suggested @azaozz and using preloaded URIs.

By tracking the already loaded dependencies on the frontend inside a single component/object, we avoid having to pass around a list of the already loaded handles, which @azaozz mentioned was a big concern in the original trac ticket discussing async script loading. How I see it is that we don't need to have something that is expressly a concern of the client's state leak into WP... I think we can completely separate these concerns between the server and the client.

@aduth that's great that there's a way to preload that information to make it accessible without needing to make another request. Totally understand the problems here. I think that given there's a need for this in the block directory, I wonder if we can do a hybridized approach.

@aduth mentioned using the block registration API for this. That sounds great to me, it's totally possible for `LazyLoad` to grab the scripts/styles handles from the block metadata, and having it in the block metadata would make it also accessible for preloading the block's assets. Then all LazyLoad needs to do is say: do I already have the list of dependency handles and URIs for this block (i.e., have they been pre-loaded)? If so (like when the block is not being added through the directory), use those. If not, then we can make some rest calls to get the dependencies.

I think that this doesn't actually add a whole lot of complexity to the `LazyLoad` component, nor do I think it would be too difficult.

Because TinyMCE is such a big win (@griffbrad puts it a nice way... that TinyMCE is the equivalent of react, react-dom, and loads more) I'd like to propose the following:
1. If y'all could tell me what it would take to get this PR merged (beyond adding tests and proper documentation) then I will get it to that point ASAP. Maybe making this experimental would help ease some concerns so folks don't start diving right into utilizing this (especially given pt 2 below)? What blockers do y'all see for me blocking this PR aside from tests and docs?
2. As a next step/fast-follow, I'll start augmenting the block registration API to handle preloading dependency URIs, as a sort of v0.2 of `LazyLoad`. This is in-line with my general goals for this project... I took this up as a good starting point for moving towards a generally async block API, TinyMCE just made sense from a perspective of how big a win it'll be for almost every single block editor user.

**Comment by azaozz** [comment]
> the REST call strategy is necessary when we don't know the assets ahead of time, as is the case with the block directory

Right. The question is "how" to do this in a good, future-proof way, not whether to use REST or not. As far as I see there are two general directions:
1. Get the new scripts URLs (using REST), then load them with some sort of scriptLoader that can preserve loading order or just plain old `document.write()` in the HTML head. This assumes script dependencies will be "partially calculated" in js (to remove already loaded scripts). Handling of "before" and "after" scripts may be somewhat problematic and needs to be built-in the scriptLoader.
2. Inject a script node in the head and pass the new "script handle" and already loaded handles in the src. Then in PHP calculate the dependencies, remove the already loaded, and output the rest as a concatenated script (that goes directly inside the `<script>` tag). Alternatively instead of injecting a script node it may be a REST request with `eval()`. Here the loading order will be preserved as the scripts are concatenated in order, including the "before" and "after". It would be a bit harder to debug sometimes as the lazy-loaded scripts may be concatenated.

A third, imho best, option would be to refactor/re-think/redo WPs Script Loader and bring it from 2005 to 2020 :) Add support for defer, async, type=module, do HTTP/2 (non-concatenated) loading, etc. I know, this is way out-of-scope here but will need doing sooner or later, and adding something "temporary" now just pushes it back.

In short, thinking that https://github.com/WordPress/gutenberg/pull/21244 and https://core.trac.wordpress.org/ticket/48654 are not quite ready yet and perhaps there is an opportunity to make this work better. 

As for loading TinyMCE's scripts on demand, perhaps outputting the needed URLs on initial page load would be sufficient for now. The same is probably true for outputting the translation strings, they aren't that big.

**Comment by azaozz** [comment]
> I would think the full dependency tree of all scripts would be quite large and not something to send over the wire (worth verifying).

As far as I see the actual dependency data is not that big. The "before" and "after" add quite a bit to it but not sure they would be needed/wanted initially. However don't think we'd need to pre-load that data as we can get "chunks" of it through REST at any time, and just subtract the already loaded scripts.

**Comment by StevenDufresne** [comment]
@aduth @saramarcondes Great stuff! 

I'll also admit that I overstated the HTTP request issues to draw out some of the mitigating strategies we have in mind. There are plenty of ways we will be able to create a progressive experience and this work is definitely the precursor, but in my opinion having some of those strategies on the board will be more beneficial than distracting.

>How I see it is that we don't need to have something that is expressly a concern of the client's state leak into WP 

Good call.

### Here's some info on the Block Directory:

**It currently works like this:** 
- User searches for "Awesome Block", we return results (blocks)
  - We have the block assets in the response object from the search.
    - This is error prone, especially since we are making assumptions about how a plugin is architectured based solely on its code. 
    - Plugin guidelines and a `block.json` file will make it easier, but the margin of error will never be 0.

- The user clicks 'Add Block'
  - Call an endpoint to install the block
  - Use the assets returned in the first call and inject them into the dom
  - 🙏 


**A better flow would be:**
- User searches for "Awesome Block", we return results (blocks) 
- The user clicks 'Add Block'
  - Call an endpoint to install the block
    - Receive handle
  - Pass the handle to a `Loader` component/function. 
    - It would have to run immediately -> The user wants to use the block immediately (based on the current experience)
  - `Loader` calls REST Endpoint, ... does its business.

The `Loader` component could easily be the `LazyLoad` component but as mentioned it needs to trigger immediately.

**Comment by saramarcondes** [comment]
In the related issue, @gziolo mentioned [that opening up an option for the classic block to be in block directory](https://github.com/WordPress/gutenberg/issues/21738#issuecomment-620987058) could potentially get the async loading capabilities. However, that would only solve the issue for when the classic block is not already used and would cause site owners to manage this preference. I think the solution that I have here where we create a way for blocks to load their dependencies even if they actually exist within a post is better, for a couple reasons:

1. It doesn't require users to opt-in or change any of their behaviors to get a significant performance benefit
2. It solves a broader range of TinyMCE performance problems (instead of just offsetting them for people not using TinyMCE, we solve it for everyone, even those using the classic block regularly)
3. It creates a way for _any_ block to follow this same pattern simply by declaring its dependencies (and the discussion above has [prompted me to think more deeply of forward solutions](https://github.com/WordPress/gutenberg/pull/21684#issuecomment-620817620))
4. It sets us up to remove the responsibility of loaded dependencies from the block directory and off-load it to the block API itself (as in, we could remove the dependency loading code from the block directory altogether as the block's dependencies would be loaded when the block is first used, whether that's immediately after the block is installed or later if we decide user behaviors aren't inline with the current dependency loading strategy).

Ultimately, I think this boils down to this: the block directory already asynchronously loads dependencies for the blocks that support it, but once you have installed a block, you're now paying the penalty for it every time you load your editor. This means that blocks have a hidden cost that isn't obvious and would be impossible to communicate. Ultimately, that means that block developers get saddled with trying to make harder performance decisions than are necessary, if they're performance minded at all.

Moving towards a solution where we could get all blocks with dependencies to asynchronously load their dependencies means that we _remove that hidden cost_. It also opens us up for nice performance wins for core blocks with big dependencies like the classic block.

@azaozz and @ellatrix, something @gziolo mentioned is that there may be some hidden dependency on TinyMCE in the editor that I haven't noticed yet. @gziolo said that y'all two would be the best to ask about this. When I search through the Gutenberg codebase, the only references to `window.tinymce` are either null safe or inside the `edit` of the classic block (which is guarded by the dependency loading wrapper). Is there anywhere else I need to take a look to make sure the core editor is okay? Smoke testing it locally I haven't noticed any issues. Also note that this work only prevents TinyMCE's default loading behavior when in the block editor, so everywhere else in the admin that uses TinyMCE won't be disrupted.
